src/high-level/matrix.lisp

```diff
@@ -0,0 +1,557 @@
+        (let ((index-function
+                (ecase (matrix-layout matrix)
+                  (:row-major #'row-major-index)
+                  (:column-major #'column-major-index)))
+              (shape (shape matrix)))
+          (loop :for row :below (matrix-nrows matrix)
+                :do (loop :for col :from row :below (matrix-ncols matrix)
+                          :do (rotatef
+                               (aref (storage matrix) (funcall index-function (list row col) shape))
+                               (aref (storage matrix) (funcall index-function (list col row) shape)))))
+          (rotatef (matrix-ncols matrix) (matrix-nrows matrix))))
+    matrix))
+
+(defgeneric transpose (matrix)
+  (:documentation "Create a new matrix containing the transpose of MATRIX")
+  (:method ((matrix matrix))
+    "Transpose a matrix by copying values.
+If fast is t then just change layout. Fast can cause problems when you want to multiply specifying transpose."
+    (let ((new-matrix (copy-tensor matrix)))
+      (setf (matrix-ncols new-matrix) (matrix-nrows matrix)
+            (matrix-nrows new-matrix) (matrix-ncols matrix))
+      (let ((index-function
+              (ecase (layout matrix)
+                (:row-major #'row-major-index)
+                (:column-major #'column-major-index)))
+            (shape (shape matrix)))
+        (loop :for row :below (matrix-nrows matrix)
+              :do (loop :for col :below (matrix-ncols matrix)
+                        :do (let ((index1 (funcall index-function (list row col) shape))
+                                  (index2 (funcall index-function (list col row) (shape new-matrix))))
+                              (setf (aref (storage new-matrix) index2) (aref (storage matrix) index1))))))
+      new-matrix)))
+
+;; TODO: allow setf on matrix diag
+(defgeneric diag (matrix)
+  (:documentation "Get a list of the diagonal elements of MATRIX")
+  (:method ((matrix matrix))
+    (policy-cond:policy-if
+     (< speed safety)
+     (assert-square-matrix matrix)
+     nil)
+    (let ((rows (nrows matrix)))
+      (loop :for i :below rows
+            :collect (tref matrix i i)))))
+
+(defgeneric trace (matrix)
+  (:documentation "Get the trace of MATRIX (sum of diagonals)")
+  (:method ((matrix matrix))
+    (policy-cond:policy-if
+     (< speed safety)
+     (assert-square-matrix matrix)
+     nil)
+    (loop :for i :below (nrows matrix)
+          :sum (tref matrix i i))))
+
+(defgeneric det (matrix)
+  (:documentation "Compute the determinant of a square matrix MATRIX")
+  (:method ((matrix matrix))
+    (policy-cond:policy-if
+     (< speed safety)
+     (assert-square-matrix matrix)
+     nil)
+    (let ((d 1))
+      (multiple-value-bind (a ipiv) (lu matrix)
+        (dotimes (i (nrows matrix))
+          (setq d (* d (tref a i i))))
+        (dotimes (i (size ipiv))
+          (unless (cl:= (1+ i) (tref ipiv i))
+            (setq d (- d))))
+        d))))
```

Comment:
Not for this PR but this is a bad idea. Better to compute an LU decomposition and multiply the diagonal elements of the U matrix.

Reply:
Isn't that what this does? Not entirely sure though because this came from the old `high-level`